File: 61.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not head or k == 0:
            return head
        slow = fast = head
        length = 0
        while fast:
            fast = fast.next
            length += 1
        k %= length
        fast = head
        for _ in range(k):
            fast = fast.next
        while fast and fast.next:
            fast = fast.next
            slow = slow.next
        fast.next = head
        head = slow.next
        slow.next = None
        return head

File: 61.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == NULL or k == 0){
            return head;
        }
        ListNode* slow = head;
        ListNode* fast = head;
        int length = 0;
        while (fast != NULL){
            length++;
            fast = fast->next;
        }
        fast = head;
        k %= length;
        for (int i = 0; i < k; i++){
            fast = fast->next;
            
        }
        while (fast and fast->next){
            fast = fast->next;
            slow = slow->next;
        }
        // slow->next = NULL;
        fast->next = head;
        head = slow->next;
        slow->next = NULL;
        return head;
        
        
    }
};
